warehouse/packaging/tasks.py

```diff
@@ -14,14 +14,73 @@
 
 from itertools import product
 
+import pip_api
+
 from google.cloud.bigquery import LoadJobConfig
 
 from warehouse import tasks
+from warehouse.accounts.models import TitanPromoCode, User
 from warehouse.cache.origin import IOriginCache
+from warehouse.email import send_two_factor_mandate_email
 from warehouse.packaging.models import Description, File, Project, Release
 from warehouse.utils import readme
 
 
+@tasks.task(ignore_result=True, acks_late=True)
+def compute_2fa_mandate(request):
+    # Get our own production dependencies
+    our_dependencies = set(
+        pip_api.parse_requirements("./requirements/main.txt")
+        | pip_api.parse_requirements("./requirements/deploy.txt")
+    )
+
+    bq = request.find_service(name="gcloud.bigquery")
+
+    # Get the top N projects in the last 6 months
+    query = bq.query(
+        """ SELECT
+              COUNT(*) AS num_downloads,
+              file.project as project_name
+            FROM
+              {table}
+            WHERE
+              DATE(timestamp) BETWEEN DATE_TRUNC(
+                DATE_SUB(CURRENT_DATE(), INTERVAL 6 MONTH), MONTH
+              )
+              AND CURRENT_DATE()
+            GROUP BY
+              file.project
+            ORDER BY
+              num_downloads DESC
+            LIMIT
+              {cohort_size}
+        """.format(
+            table=request.registry.settings["warehouse.downloads_table"],
+            cohort_size=request.registry.settings[
+                "warehouse.two_factor_mandate.cohort_size"
+            ],
```

Comment:
I don't think there's any possibility for an injection vulnerability here, but just as a matter of best practice: maybe parametrized queries here instead of raw interpolation?

Looks like biquery supports parametrization with `@identifier`: https://cloud.google.com/bigquery/docs/parameterized-queries

Reply:
This is a good tip, but I'm not concerned about it here, as these values only set by PyPI admins.